core/src/main/kotlin/com/justai/jaicf/model/activation/ActivationSelector.kt

```diff
@@ -19,29 +20,44 @@ abstract class ActivationSelector {
      * @see com.justai.jaicf.activator.selection.ActivationByConfidence
      * @see com.justai.jaicf.activator.selection.ActivationByContextPenalty
      */
-    abstract fun selectActivation(botContext: BotContext, activations: List<Activation>): Activation
+    fun selectActivation(botContext: BotContext, activations: List<Activation>): Activation
+}
 
-    /**
-     * Sorts activations by context difference - number of transitions between current state and activation target state.
-     * */
-    protected fun Collection<Activation>.sortedByContext(currentState: StatePath) = sortedBy { activation ->
-        activation.state?.let { targetState ->
-            calculateStatesDifference(StatePath.parse(targetState), currentState)
-        }
+/**
+ * Sorts list of activation by relevance to currentContext - first child nodes, then siblings, then parents, etc.
+ * */
+fun Collection<Activation>.sortedByContext(currentContext: String): List<Activation> {
+    val (fromRoot, onPath) = partition {
+        val isRoot = it.state!!.commonPrefixWith(currentContext) == "/"
+        val isIndirectChild = it.state.asStatePath().isIndirectChildOf(currentContext.asStatePath())
```

Comment:
What if an indirect child comes not from the root, but from some of our ancestors? It should have a bigger priority, shouldn't it?
current = `/a/b`
indirect child = `/a/b/c/d`, fromState = `/a`

Reply:
At this point we don't know if indirect child was be activated by fromState or by global activator, and this issue cannot be solved with current architecture.

Looks like `ActivationSelector` should start to know about Transitions, should generate sorted transitions with or without penalty. Architectual rework will be done in further commits.

Reply:
So, we are about to break our current logic (that is, by the way, considered as default in this PR) in order to add some other implementation. I don't think it should be done this way.

Reply:
No, I meant that we shouldn't merge this PR until I've made changes, so we wont break current logic

Reply:
Ah okay, I've misunderstood you, sorry.
And about your previous message. I don't think `ActivationSelector` should be aware of transitions or generate them on its own. We should provide all required data through an `Activation` object, precisely through some state-like object in it.

Reply:
I think at some point some entity should take transition management from `Activator`, i thought about changing this right now, but it leads to way to many changes in architecture.

I think i'll implement your idea with providing `Activation` with necessary data. Thanks for idea!

Reply:
@nikvoloshin I've added StatesTransition class holding information about transition (fromState, toState, distance). 

But there is much to improve in future. The problems I see now are:
1. checkStrictTransitions() in BotEngine class. It should not be there, also Strict Transition should be a separate entity.
2. Distance should be calculated when we create transition (generateTransitions).
3. I don't like that we have two classes (Transition and StatesTransition (or ActivationTransition?)) containing mostly the same data. Looks like we need only one of them.

But solving any of these problems requires a number of changes unrelated to issue. I'd like to solve them in separate PR in future.

Reply:
@Denire Thanks for the changes! I like that now `Activation` has more info about states and transitions. 
One thing to discuss: it seems like we don't really need `StatesTransition` entity. We can pass the `Transition` object to the `Activation` and provide all utility functionality as extension functions because `StatesTransition` is a stateless and immutable object with `fromState` and `toState` as primary parameters.
So, please think in this direction, maybe it'd be better just to provide utility extension functions on `State` and `Transition` instead of creating utility-like entity

Reply:
@nikvoloshin I thought about it and found two reasons not to use `Transition` object:

1. Transition has `ActivationRule` required property, and in case of Slot Filling or Strict Activations it will require unpleasant changes in BotEngine and other places, which don't care about `ActivationRule`. Their job is only to create and object and pass it further. 
Making `ActivationRule` optional also doesn't look good.

2. It may be better to use Transition for ScenarioBuilding and StatesTransition (ActivationTransition?) for processing activations. Therefore we have independent objects, and in case we need to modify one of them, we won't be affecting many classes, which use only ActivationTransition logic. 

upd: I'll rename StatesTransition to ActivationTransition.